src/envoy/tcp/alpn_proxy/alpn_proxy.cc

```diff
@@ -0,0 +1,234 @@
+  if ((connection_protocol_read_ && !alpn_protocol_found_) ||
+      (alpn_protocol_found_ && read_done_)) {
+    return Network::FilterStatus::Continue;
+  }
+  if (read_callbacks_->connection().nextProtocol() != config_->protocol_) {
+    connection_protocol_read_ = true;
+    config_->stats().alpn_protocol_not_found_.inc();
+    return Network::FilterStatus::Continue;
+  }
+  if (!alpn_protocol_found_) {
+    config_->stats().alpn_protocol_found_.inc();
+    alpn_protocol_found_ = true;
+  }
+
+  // TODO(gargnupur): Try to move this just after alpn protocol is determined
+  // and first onData is called in Downstream filter.
+  if (config_->filter_direction_ ==
+      envoy::tcp::alpnproxy::config::AlpnProxy_FilterDirection::
+          AlpnProxy_FilterDirection_DOWNSTREAM) {
+    writeNodeMetadata();
+  }
+
+  tryReadInitialProxyHeader(data);
+  if (conn_state_ == NeedMoreDataIinitialHeader) {
+    return Network::FilterStatus::StopIteration;
+  }
+  if (conn_state_ == Invalid) {
+    read_done_ = true;
+    return Network::FilterStatus::Continue;
+  }
+  tryReadProxyHeaderProto(data);
+  if (conn_state_ == NeedMoreDataProxyHeader) {
+    return Network::FilterStatus::StopIteration;
+  }
+
+  read_done_ = true;
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus AlpnProxyFilter::onNewConnection() {
+  return Network::FilterStatus::Continue;
+}
+
+Network::FilterStatus AlpnProxyFilter::onWrite(Buffer::Instance&, bool) {
+  // No work needed if we have read connection protocol and its not the expected
+  // alpn protocol or if alpn protocol is found but we are still writing data.
+  if ((connection_protocol_read_ && !alpn_protocol_found_) ||
+      (alpn_protocol_found_ && write_done_)) {
+    return Network::FilterStatus::Continue;
+  }
+  if (read_callbacks_->connection().nextProtocol() != config_->protocol_) {
+    connection_protocol_read_ = true;
+    config_->stats().alpn_protocol_not_found_.inc();
+    return Network::FilterStatus::Continue;
+  }
+  if (!alpn_protocol_found_) {
+    config_->stats().alpn_protocol_found_.inc();
+    alpn_protocol_found_ = true;
+  }
+
+  // TODO(gargnupur): Try to move this just after alpn protocol is determined
+  // and first onWrite is called in Upstream filter.
+  if (config_->filter_direction_ ==
+      envoy::tcp::alpnproxy::config::AlpnProxy_FilterDirection::
+          AlpnProxy_FilterDirection_UPSTREAM) {
+    writeNodeMetadata();
+  }
+  write_done_ = true;
+  return Network::FilterStatus::Continue;
+}
+
+void AlpnProxyFilter::writeNodeMetadata() {
+  if (conn_state_ != WriteMetadata) {
+    return;
+  }
+
+  if (local_info_.node().has_metadata()) {
+    auto metadata_fields = local_info_.node().metadata().fields();
+    auto node_metadata = metadata_fields.find(config_->node_metadata_id_);
+    if (node_metadata != metadata_fields.end()) {
+      std::unique_ptr<::Envoy::Buffer::OwnedImpl> buf =
+          constructProxyHeaderData(node_metadata->second.struct_value());
```

Comment:
Heads up that we are switching to using filter state instead of metadata. Therefore, it might be worth encapsulating all code that has to do with metadata reading and saving into functions.

Reply:
Thanks for the heads up! Moved into readMetadata and writeMetadata functions